benchmark/src/DriverBench/ParallelMultiFileExportBench.php

```diff
@@ -0,0 +1,225 @@
+                    'skip' => 5_000 * $params['chunk'] * $i,
+                ]);
+
+                // Exit the child process
+                exit(0);
+            }
+
+            if ($pid === -1) {
+                throw new RuntimeException('Failed to fork');
+            }
+
+            // Keep the forked process id to wait for it later
+            $pids[$pid] = true;
+        }
+
+        // Wait for all child processes to finish
+        while ($pids !== []) {
+            $pid = pcntl_waitpid(-1, $status);
+            unset($pids[$pid]);
+        }
+    }
+
+    /**
+     * Using amphp/parallel with worker pool
+     *
+     * @param array{chunk:int} $params
+     */
+    #[ParamProviders(['provideChunkParams'])]
+    public function benchAmpWorkers(array $params): void
+    {
+        $workerPool = new ContextWorkerPool(ceil(100 / $params['chunk']), new ContextWorkerFactory());
+
+        $futures = [];
+        foreach (array_chunk(self::getFileNames(), $params['chunk']) as $i => $files) {
+            $futures[] = $workerPool->submit(
+                new ExportFileTask(
+                    files: $files,
+                    options: [
+                        'limit' => 5_000 * $params['chunk'],
+                        'skip' => 5_000 * $params['chunk'] * $i,
+                    ],
+                ),
+            )->getFuture();
+        }
+
+        foreach (Future::iterate($futures) as $future) {
+            $future->await();
+        }
+    }
+
+    public static function provideChunkParams(): Generator
+    {
+        yield 'by 1' => ['chunk' => 1];
+        yield 'by 2' => ['chunk' => 2];
+        yield 'by 4' => ['chunk' => 4];
+        yield 'by 8' => ['chunk' => 8];
+        yield 'by 13' => ['chunk' => 13];
+        yield 'by 20' => ['chunk' => 20];
+        yield 'by 100' => ['chunk' => 100];
+    }
+
+    /**
+     * Export a query to a file
+     */
+    public static function exportFile(array|string $files, array $filter = [], array $options = []): void
+    {
+        $options += [
+            // bson typemap is faster on query result, but slower to JSON encode
+            'typeMap' => ['root' => 'array'],
+            // Excludes _id field to be identical to fixtures data
+            'projection' => ['_id' => 0],
+            'sort' => ['_id' => 1],
+        ];
+        $cursor = Utils::getCollection()->find($filter, $options);
+        $cursor->rewind();
+
+        foreach ((array) $files as $file) {
+            // Aggregate file in memory to reduce filesystem operations
+            $data = '';
+            for ($i = 0; $i < 5_000; $i++) {
+                $document = $cursor->current();
+                // Cursor exhausted
+                if (! $document) {
+                    break;
+                }
+
+                // We don't use MongoDB\BSON\Document::toCanonicalExtendedJSON() because
+                // it is slower than json_encode() on an array.
```

Comment:
Note that `json_encode` does not produce the same result as `toCanonicalExtendedJSON`. That said, it does produce the result we wanted.

I also had to do a double-take on this, so I wrote a small benchmark:
```
    benchToJSONViaToCanonicalExtendedJSON...I2 - Mo48.200μs (±0.00%)
    benchToJSONViaToRelaxedExtendedJSON.....I2 - Mo48.591μs (±6.09%)
    benchToJSONViaJsonEncode................I2 - Mo12.410μs (±3.55%)
```

The last benchmark actually calls `json_encode($document->toPHP(['root' => 'array']))`, so I'm a bit surprised that is faster. This might provide an opportunity to revisit the JSON serialisation logic in libbson.

Reply:
Tracked in [PHPC-2299](https://jira.mongodb.org/browse/PHPC-2299)